### ISSUE TITLE
[Bug]: Use another mod which contains HamonyPatch together with UE, causing exception

### ISSUE THREAD

**Issue by lxymahatma** [yukieiji/UnityExplorer#9]
### Are you on the latest version of UnityExplorer?

- [X] Yes, I'm on the latest version of UnityExplorer.

### Which release are you using?

MelonLoader IL2CPP CoreCLR

### Which game did this occur on?

Muse Dash
Unity Version: 2019.4.32f1

### Describe the issue.

When only have UE alone, everything is fine.
If add another mod with nothing in it (or normal methods) like the image below, together with UE, still fine
![image](https://github.com/yukieiji/UnityExplorer/assets/66151783/86ae77bf-06be-484e-8ffc-7bc75ba820c4)

However, if add a HarmonyPatch in the mod, shown below, UE will throw an exception about Il2CppRewired_Core
![image](https://github.com/yukieiji/UnityExplorer/assets/66151783/090c6351-4c98-44e1-a225-6d9cbd3fdbb2)

Detailed information is in the log output

### Relevant log output

```shell
[18:18:14.273] ------------------------------
[18:18:14.299] MelonLoader v0.6.1 Open-Beta
[18:18:14.300] OS: Windows 11
[18:18:14.301] Hash Code: 696B58144EEC73C72FA36A811CA73312EEE4EC28AD5E3527BDA1E14F388A
[18:18:14.301] ------------------------------
[18:18:14.302] Game Type: Il2cpp
[18:18:14.302] Game Arch: x64
[18:18:14.302] ------------------------------
[18:18:14.302] Core::BasePath = C:\Program Files (x86)\Steam\steamapps\common\Muse Dash
[18:18:14.303] Game::BasePath = C:\Program Files (x86)\Steam\steamapps\common\Muse Dash
[18:18:14.303] Game::DataPath = C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\musedash_Data
[18:18:14.303] Game::ApplicationPath = C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\musedash.exe
[18:18:14.304] Runtime Type: net6
[18:18:14.462] ------------------------------
[18:18:14.463] Game Name: MuseDash
[18:18:14.463] Game Developer: PeroPeroGames
[18:18:14.465] Unity Version: 2019.4.32f1
[18:18:14.465] Game Version: 3.12.1
[18:18:14.466] ------------------------------

[18:18:14.955] Preferences Loaded!

[18:18:14.972] Loading Plugins from 'C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\Plugins'...
[18:18:14.974] 0 Plugins loaded.
[18:18:15.469] [MelonStartScreen] Initializing...
[18:18:15.489] [MelonStartScreen] Using Start Screen Theme: "MDMC"
[18:18:15.992] Loading Il2CppAssemblyGenerator...
[18:18:16.006] [Il2CppAssemblyGenerator] Contacting RemoteAPI...
[18:18:16.394] [Il2CppAssemblyGenerator] RemoteAPI.DumperVersion = null
[18:18:16.396] [Il2CppAssemblyGenerator] RemoteAPI.ObfuscationRegex = null
[18:18:16.400] [Il2CppAssemblyGenerator] RemoteAPI.MappingURL = null
[18:18:16.403] [Il2CppAssemblyGenerator] RemoteAPI.MappingFileSHA512 = null
[18:18:16.416] [Il2CppAssemblyGenerator] Using Dumper Version: 2022.1.0-pre-release.10
[18:18:16.418] [Il2CppAssemblyGenerator] Using Il2CppInterop Version = 1.4.5
[18:18:16.425] [Il2CppAssemblyGenerator] Using Unity Dependencies Version = 2019.4.32
[18:18:16.430] [Il2CppAssemblyGenerator] Using Deobfuscation Regex = null
[18:18:16.437] [Il2CppAssemblyGenerator] Cpp2IL is up to date.
[18:18:16.444] [Il2CppAssemblyGenerator] UnityDependencies is up to date.
[18:18:16.449] [Il2CppAssemblyGenerator] Checking GameAssembly...
[18:18:16.623] [Il2CppAssemblyGenerator] Assembly is up to date. No Generation Needed.

[18:18:16.642] Loading Mods from 'C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\Mods'...
[18:18:16.643] ------------------------------
[18:18:16.649] Melon Assembly loaded: '.\Mods\TestMod.dll'
[18:18:16.650] SHA256 Hash: '59f92617bf8d5dd10632eea95dfaecced11688578fbab41bafd482d78596a461'
[18:18:16.683] Melon Assembly loaded: '.\Mods\UnityExplorer.ML.IL2CPP.CoreCLR.dll'
[18:18:16.685] SHA256 Hash: '2d2493c3febf25f387c29555a813cc2c1ca75bb4d52fe83bfd1ff20a91330436'

[18:18:16.998] ------------------------------
[18:18:16.998] TestMod v1.0.0
[18:18:16.998] by test
[18:18:16.999] Assembly: TestMod.dll
[18:18:17.000] ------------------------------
[18:18:17.008] ------------------------------
[18:18:17.008] UnityExplorer v4.11.0
[18:18:17.008] by Sinai, yukieiji
[18:18:17.009] Assembly: UnityExplorer.ML.IL2CPP.CoreCLR.dll
[18:18:17.009] ------------------------------
[18:18:17.010] ------------------------------
[18:18:17.013] 2 Mods loaded.

[18:18:17.414] Registered mono type Il2CppInterop.Runtime.DelegateSupport+Il2CppToMonoDelegateReference in il2cpp domain
[18:18:17.439] Registered mono type MelonLoader.Support.MonoEnumeratorWrapper in il2cpp domain
[18:18:17.441] Registered mono type MelonLoader.Support.SM_Component in il2cpp domain
[18:18:17.448] Support Module Loaded: C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\MelonLoader\Dependencies\SupportModules\Il2Cpp.dll
[18:18:17.489] [UnityExplorer] UnityExplorer 4.11.0 initializing...
[18:18:17.507] [UnityExplorer] [UniverseLib] UniverseLib 1.5.4 initializing...
[18:18:17.510] Registered mono type UniverseLib.UniversalBehaviour in il2cpp domain
[18:18:17.937] [UnityExplorer] System.BadImageFormatException: Could not load file or assembly 'Il2CppRewired_Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. Attempted to load a program with an incorrect format.
File name: 'Il2CppRewired_Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' ---> System.BadImageFormatException: Format of the executable (.exe) or library (.dll) is invalid.
   at MelonLoader.Fixes.DotnetAssemblyLoadContextFix.PreAlcLoadFromPath(String ilPath) in D:\a\MelonLoader\MelonLoader\MelonLoader\Fixes\DotnetAssemblyLoadContextFix.cs:line 91
   at NativeDetour_Wrapper<System.Runtime.Loader.AssemblyLoadContext::LoadFromPath>?0(IntPtr ptrNativeAssemblyLoadContext, String ilPath, String niPath, ObjectHandleOnStack retAssembly)
   at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyPath(String assemblyPath)
   at MelonLoader.Fixes.DotnetLoadFromManagedFolderFix.TryLoad(AssemblyLoadContext alc, String path) in D:\a\MelonLoader\MelonLoader\MelonLoader\Fixes\DotnetLoadFromManagedFolderFix.cs:line 32
   at MelonLoader.Fixes.DotnetLoadFromManagedFolderFix.OnResolve(AssemblyLoadContext alc, AssemblyName name) in D:\a\MelonLoader\MelonLoader\MelonLoader\Fixes\DotnetLoadFromManagedFolderFix.cs:line 53
   at System.Runtime.Loader.AssemblyLoadContext.GetFirstResolvedAssemblyFromResolvingEvent(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingEvent(AssemblyName assemblyName)
   at System.Runtime.Loader.AssemblyLoadContext.ResolveUsingResolvingEvent(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
   at System.RuntimeTypeHandle.GetDeclaringType(RuntimeType type)
   at System.RuntimeType.RuntimeTypeCache.GetEnclosingType()
   at System.RuntimeType.RuntimeTypeCache.GetNameSpace()
   at System.RuntimeType.get_Namespace()
   at UniverseLib.ReflectionUtility.CacheTypes(Assembly asm)
   at UniverseLib.ReflectionUtility.SetupTypeCache()
   at UniverseLib.ReflectionUtility.Initialize()
   at UniverseLib.Il2CppReflection.Initialize()
   at UniverseLib.ReflectionUtility.Init()
   at UniverseLib.Universe.Init(Single startupDelay, Action onInitialized, Action`2 logHandler, UniverseLibConfig config)
   at UnityExplorer.ExplorerCore.Init(IExplorerLoader loader)
   at UnityExplorer.ExplorerMelonMod.OnApplicationStart()
   at MelonLoader.MelonEvent.<>c.<Invoke>b__1_0(LemonAction x) in D:\a\MelonLoader\MelonLoader\MelonLoader\Melons\Events\MelonEvent.cs:line 174
   at MelonLoader.MelonEventBase`1.Invoke(Action`1 delegateInvoker) in D:\a\MelonLoader\MelonLoader\MelonLoader\Melons\Events\MelonEvent.cs:line 143
```

**Comment by yukieiji** [comment]
I would like to confirm the following three points.
1. can you send me the logs and screenshots of how it works with UE only?
2. do other mods besides UE work correctly?(with harmony patch)
3. is it the same with ML 0.6.2?

**Comment by lxymahatma** [comment]
> I would like to confirm the following three points.
> 
> 1. can you send me the logs and screenshots of how it works with UE only?
> 2. do other mods besides UE work correctly?(with harmony patch)
> 3. is it the same with ML 0.6.2?

1. [20:44:27.926] ------------------------------
[20:44:27.951] MelonLoader v0.6.1 Open-Beta
[20:44:27.952] OS: Windows 11
[20:44:27.952] Hash Code: 696B58144EEC73C72FA36A811CA73312EEE4EC28AD5E3527BDA1E14F388A
[20:44:27.952] ------------------------------
[20:44:27.953] Game Type: Il2cpp
[20:44:27.953] Game Arch: x64
[20:44:27.953] ------------------------------
[20:44:27.953] Core::BasePath = C:\Program Files (x86)\Steam\steamapps\common\Muse Dash
[20:44:27.953] Game::BasePath = C:\Program Files (x86)\Steam\steamapps\common\Muse Dash
[20:44:27.953] Game::DataPath = C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\musedash_Data
[20:44:27.953] Game::ApplicationPath = C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\musedash.exe
[20:44:27.954] Runtime Type: net6
[20:44:28.108] ------------------------------
[20:44:28.109] Game Name: MuseDash
[20:44:28.109] Game Developer: PeroPeroGames
[20:44:28.111] Unity Version: 2019.4.32f1
[20:44:28.111] Game Version: 3.12.1
[20:44:28.111] ------------------------------

[20:44:28.534] Preferences Loaded!

[20:44:28.547] Loading Plugins from 'C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\Plugins'...
[20:44:28.548] 0 Plugins loaded.
[20:44:29.061] [MelonStartScreen] Initializing...
[20:44:29.078] [MelonStartScreen] Using Start Screen Theme: "MDMC"
[20:44:29.563] Loading Il2CppAssemblyGenerator...
[20:44:29.582] [Il2CppAssemblyGenerator] Contacting RemoteAPI...
[20:44:30.002] [Il2CppAssemblyGenerator] RemoteAPI.DumperVersion = null
[20:44:30.004] [Il2CppAssemblyGenerator] RemoteAPI.ObfuscationRegex = null
[20:44:30.007] [Il2CppAssemblyGenerator] RemoteAPI.MappingURL = null
[20:44:30.012] [Il2CppAssemblyGenerator] RemoteAPI.MappingFileSHA512 = null
[20:44:30.025] [Il2CppAssemblyGenerator] Using Dumper Version: 2022.1.0-pre-release.10
[20:44:30.027] [Il2CppAssemblyGenerator] Using Il2CppInterop Version = 1.4.5
[20:44:30.029] [Il2CppAssemblyGenerator] Using Unity Dependencies Version = 2019.4.32
[20:44:30.031] [Il2CppAssemblyGenerator] Using Deobfuscation Regex = null
[20:44:30.036] [Il2CppAssemblyGenerator] Cpp2IL is up to date.
[20:44:30.041] [Il2CppAssemblyGenerator] UnityDependencies is up to date.
[20:44:30.044] [Il2CppAssemblyGenerator] Checking GameAssembly...
[20:44:30.182] [Il2CppAssemblyGenerator] Assembly is up to date. No Generation Needed.

[20:44:30.213] Loading Mods from 'C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\Mods'...
[20:44:30.215] ------------------------------
[20:44:30.256] Melon Assembly loaded: '.\Mods\UnityExplorer.ML.IL2CPP.CoreCLR.dll'
[20:44:30.256] SHA256 Hash: '2d2493c3febf25f387c29555a813cc2c1ca75bb4d52fe83bfd1ff20a91330436'

[20:44:30.418] ------------------------------
[20:44:30.419] UnityExplorer v4.11.0
[20:44:30.419] by Sinai, yukieiji
[20:44:30.420] Assembly: UnityExplorer.ML.IL2CPP.CoreCLR.dll
[20:44:30.420] ------------------------------
[20:44:30.421] ------------------------------
[20:44:30.421] 1 Mod loaded.

[20:44:30.825] Registered mono type Il2CppInterop.Runtime.DelegateSupport+Il2CppToMonoDelegateReference in il2cpp domain
[20:44:30.853] Registered mono type MelonLoader.Support.MonoEnumeratorWrapper in il2cpp domain
[20:44:30.855] Registered mono type MelonLoader.Support.SM_Component in il2cpp domain
[20:44:30.861] Support Module Loaded: C:\Program Files (x86)\Steam\steamapps\common\Muse Dash\MelonLoader\Dependencies\SupportModules\Il2Cpp.dll
[20:44:30.896] [UnityExplorer] UnityExplorer 4.11.0 initializing...
[20:44:30.918] [UnityExplorer] [UniverseLib] UniverseLib 1.5.4 initializing...
[20:44:30.922] Registered mono type UniverseLib.UniversalBehaviour in il2cpp domain
[20:44:31.138] Registered mono type UniverseLib.Runtime.Il2Cpp.Il2CppManagedEnumerator in il2cpp domain
[20:44:31.183] [UnityExplorer] [UniverseLib] Finished UniverseLib initial setup.
[20:44:31.191] Registered mono type UnityExplorer.ExplorerBehaviour in il2cpp domain
[20:44:31.248] [UnityExplorer] Initialized UnityCrashPrevention for: Canvas.get_renderingDisplaySize
[20:44:33.229] [UnityExplorer] [UniverseLib] Loaded Unhollowed modules in 2.089 seconds.
[20:44:33.347] [UnityExplorer] [UniverseLib] Setup deobfuscation cache in 0.117000006 seconds.
[20:44:33.538] [UnityExplorer] [UniverseLib] Initialized Legacy Input support
[20:44:33.576] [UnityExplorer] [UniverseLib] 	 IL2CPP method has no corresponding pointer, aborting patch of UnityEngine.AssetBundle.UnloadAllAssetBundles
[20:44:33.578] Registered mono type UniverseLib.AssetBundle in il2cpp domain
[20:44:33.646] [UnityExplorer] [UniverseLib] Loaded modern bundle for Unity 2019.4.32f1
[20:44:33.663] [UnityExplorer] [UniverseLib] UniverseLib 1.5.4 initialized.
[20:44:33.664] [UnityExplorer] Creating UI...
[20:44:34.102] [UnityExplorer] UnityExplorer 4.11.0 (IL2CPP) initialized.
[20:44:50.497] Preferences Saved!

2. Yes, the other mods are working correctly

3. Yes, the same error

**Comment by yukieiji** [comment]
OK

Will this work?
[UnityExplorer.MelonLoader.IL2CPP.CoreCLR_tryfix.zip](https://github.com/yukieiji/UnityExplorer/files/14111991/UnityExplorer.MelonLoader.IL2CPP.CoreCLR.zip)

**Comment by lxymahatma** [comment]
> OK
> 
> Will this work? [UnityExplorer.MelonLoader.IL2CPP.CoreCLR_tryfix.zip](https://github.com/yukieiji/UnityExplorer/files/14111991/UnityExplorer.MelonLoader.IL2CPP.CoreCLR.zip)

hmm no, still the same issue
[Latest.log](https://github.com/yukieiji/UnityExplorer/files/14112047/Latest.log)

**Comment by yukieiji** [comment]
Oops, forgot to rebuild UniverseLib... 
Here is the rebuild, does it work?
[UnityExplorer.MelonLoader.IL2CPP.CoreCLR_UniverseLibFix.zip](https://github.com/yukieiji/UnityExplorer/files/14112115/UnityExplorer.MelonLoader.IL2CPP.CoreCLR.zip)

**Comment by lxymahatma** [comment]
> Oops, forgot to rebuild UniverseLib... Here is the rebuild, does it work? [UnityExplorer.MelonLoader.IL2CPP.CoreCLR_UniverseLibFix.zip](https://github.com/yukieiji/UnityExplorer/files/14112115/UnityExplorer.MelonLoader.IL2CPP.CoreCLR.zip)

[Latest.log](https://github.com/yukieiji/UnityExplorer/files/14112131/Latest.log)

Yeah it's working ~~tho UE still throw out errors but it loads correctly ig~~
![image](https://github.com/yukieiji/UnityExplorer/assets/66151783/bcf172ad-f082-4362-bce6-7f8a590c8781)

**Comment by yukieiji** [comment]
This bug and the error displayed now are caused by the inability to load the DLL when creating the namespace cache.
Probably due to a bug around Il2Cpp that prevented the dll from dumping properly...

**Comment by lxymahatma** [comment]
> This bug and the error displayed now are caused by the inability to load the DLL when creating the namespace cache. Probably due to a bug around Il2Cpp that prevented the dll from dumping properly...

ah i see, so it's likely the melonloader side problem(?

**Comment by yukieiji** [comment]
I think it's a problem due to the combination of il2CppInterop and ML.
It's been almost a year since other UEs have taken similar action, so it seems like a difficult problem to solve.

**Comment by yukieiji** [comment]
fixed in [v4.11.1](https://github.com/yukieiji/UnityExplorer/releases/tag/v4.11.1)
